src/agent/discovery_proxy.cpp

```diff
@@ -0,0 +1,369 @@
+
+        subscription.mSubscriptionCount--;
+        assert(subscription.mSubscriptionCount >= 0);
+
+        if (subscription.mSubscriptionCount == 0)
+        {
+            mSubscriptions.erase(it);
+        }
+
+        otbrLog(OTBR_LOG_DEBUG, "[dnssd discovery proxy] service subscriptions: %sx%d", it->ToString().c_str(),
+                it->mSubscriptionCount);
+
+        if (GetServiceSubscriptionCount(instanceName, serviceName) == 0)
+        {
+            if (subscription.mHostName.empty())
+            {
+                mMdnsPublisher.UnsubscribeService(serviceName, instanceName);
+            }
+            else
+            {
+                mMdnsPublisher.UnsubscribeHost(hostName);
+            }
+        }
+    }
+exit:
+    if (error != OTBR_ERROR_NONE)
+    {
+        otbrLog(OTBR_LOG_WARNING, "[dnssd discovery proxy] failed to unsubscribe %s.%s: %s",
+                instanceName.empty() ? "ANY" : instanceName.c_str(), aFullName, otbrErrorString(error));
+    }
+}
+
+void DiscoveryProxy::OnServiceDiscovered(const std::string &                            aType,
+                                         const Mdns::Publisher::DiscoveredInstanceInfo &aInstanceInfo)
+{
+    otDnssdServiceInstanceInfo instanceInfo;
+
+    otbrLog(OTBR_LOG_INFO,
+            "[dnssd discovery proxy] service discovered: %s, instance %s hostname %s address %s:%d priority %d "
+            "weight %d",
+            aType.c_str(), aInstanceInfo.mName.c_str(), aInstanceInfo.mHostName.c_str(),
+            aInstanceInfo.mAddress.ToString().c_str(), aInstanceInfo.mPort, aInstanceInfo.mPriority,
+            aInstanceInfo.mWeight);
+
+    CheckServiceNameSanity(aType);
+    CheckHostnameSanity(aInstanceInfo.mHostName);
+
+    instanceInfo.mAddress   = *reinterpret_cast<const otIp6Address *>(&aInstanceInfo.mAddress);
+    instanceInfo.mPort      = aInstanceInfo.mPort;
+    instanceInfo.mPriority  = aInstanceInfo.mPriority;
+    instanceInfo.mWeight    = aInstanceInfo.mWeight;
+    instanceInfo.mTxtLength = static_cast<uint16_t>(aInstanceInfo.mTxtData.size());
+    instanceInfo.mTxtData   = aInstanceInfo.mTxtData.data();
+    instanceInfo.mTtl       = std::min(aInstanceInfo.mTtl, static_cast<uint32_t>(kServiceTtlCapLimit));
+
+    std::for_each(mSubscriptions.begin(), mSubscriptions.end(), [&](const MdnsSubscription &aSubscription) {
+        if (aSubscription.MatchesServiceInstance(aType, aInstanceInfo.mName))
+        {
+            std::string serviceFullName  = aType + "." + aSubscription.mDomain;
+            std::string hostName         = TranslateDomain(aInstanceInfo.mHostName, aSubscription.mDomain);
+            std::string instanceFullName = aInstanceInfo.mName + "." + serviceFullName;
+
+            instanceInfo.mFullName = instanceFullName.c_str();
+            instanceInfo.mHostName = hostName.c_str();
+
+            otDnssdQueryNotifyServiceInstance(mNcp.GetInstance(), serviceFullName.c_str(), &instanceInfo);
```

Comment:
The user could subscribe to a specific service instance, right? Shouldn't we notify OT with the full service instance name  in this case?

Reply:
The full service instance name is in `instanceInfo` already

Reply:
Technically `serviceFullName` is redundant info because `instanceInfo` already contains the full instance name. However, we notify OT core with both service full name and instance full name that OT core don't need to parse or split.

Reply:
Why not have a `mServiceName` pointer which points to middle of `mFullName` of `instanceInfo`. I believe the user will expect that the same name should be used in the notify callback.